src/base/kaldi-error.cc

```diff
@@ -177,27 +144,104 @@ MessageLogger::MessageLogger(LogMessageEnvelope::Severity severity,
   envelope_.line = line;
 }
 
+
 MessageLogger::~MessageLogger() KALDI_NOEXCEPT(false) {
+  // remove trailing '\n',
   std::string str = ss_.str();
   while (!str.empty() && str[str.length() - 1] == '\n')
     str.resize(str.length() - 1);
-  SendToLog(envelope_, str.c_str());
 
-  if (envelope_.severity > LogMessageEnvelope::Error)
-    return;
+  // print the mesage (or send to logging handler),
+  MessageLogger::SendToLog(envelope_, str.c_str());
+}
 
-  // On error, throw an exception with the message, plus traceback info if
-  // available.
-  if (!std::uncaught_exception()) {
-#ifdef HAVE_EXECINFO_H
-    throw std::runtime_error(str + "\n\n[stack trace: ]\n" +
-                             KaldiGetStackTrace() + "\n");
-#else
-    throw std::runtime_error(str);
-#endif
+
+void MessageLogger::SendToLog(const LogMessageEnvelope &envelope,
```

Comment:
please make the line after this one line up (whitespace issue).
Let's rename this function to, say, HandleMessage, because it does more than just send the message to the log.

The behavior needs to be as follows:
   - kError should lead to the error message being printed and sent to log, and then throw an exception.  The exception can have an empty message.  This important in order to get the expected behavior in multi-threaded cases where the exception is not caught, and in non-Kaldi binaries that call kaldi code.
   - kAssertFailure should lead to the above, but abort not throw an exception.
After this we can just get rid of the IsKaldiError, and instead test for whether the string is empty before printing it followed by a newline, or from a macro like KALDI_LOG.  (No point printing just a newline).